### PR TITLE
Set environment variables through the workspace

### PR DESCRIPTION
**Pull request by emillon**
This is a first step towards #205.

The idea is that we add some extra environment variables to `context.env` in `Context.create`. These are sourced from `(env)`.

Outstanding problems for now:
- "env" is not a great name for the field (the full syntax is `(env (_ (env (k v))))`. `environment_variables` is precise but too long. Maybe `env_vars` could work?
- at the moment, the env inheritance is done at the `Super_context` level, but we don't have a super-context when doing `dune exec`. Shall I move `Env_node` out of `Super_context` so that we can use that inheritance in `Context` as well?
- in `env_nodes`, there is a `workspace` component and a `context` component. `workspace` corresponds to what is set in `dune-workspace`, but what is the `context` one?

Thanks!

### CODE REVIEW

**Comment by emillon** [comment]
What do you mean by dependency tracking at the env level?

File: src/context.ml [hunk 1]
```diff
@@ -23,6 +23,20 @@ module Env_nodes = struct
...
+  let extra_vars ~profile env_nodes =
+    let make_map = function
+      | None ->
+        Env.Map.empty
+      | Some stanza ->
+        match Dune_env.Stanza.find stanza ~profile with
+        | None -> Env.Map.empty
+        | Some {env; _} -> Env.Map.of_list_exn env
+    in
+    Env.Map.union
```
Comment (rgrinberg) [inline src/context.ml]: Isn't this just `Env.Map.superpose`?

File: src/context.ml [hunk 2]
```diff
@@ -23,6 +23,20 @@ module Env_nodes = struct
...
     }
+
+  let extra_vars ~profile env_nodes =
+    let make_map = function
+      | None ->
+        Env.Map.empty
+      | Some stanza ->
+        match Dune_env.Stanza.find stanza ~profile with
+        | None -> Env.Map.empty
+        | Some {env; _} -> Env.Map.of_list_exn env
```
Comment (rgrinberg) [inline src/context.ml]: This is more a matter of taste, but I prefer to write this code monadically, and only get `Option.value ~default:Env.Map.empty` at the end. So something like:

```
let make_map stanza =
  Option.value ~default:Env.Map.empty (
    stanza >>= Dune_env.Stanza.find stanza ~profile >>| fun {env; _} -> Env.Map.of_list_exn env
  )
```

**Comment by rgrinberg** [comment]
> What do you mean by dependency tracking at the env level?

Just that changing the environment in the env stanza will trigger a rebuild.

**Comment by rgrinberg** [comment]
> "env" is not a great name for the field (the full syntax is (env (_ (env (k v)))). environment_variables is precise but too long. Maybe env_vars could work?

What about `exec-env`?

**Comment by emillon** [comment]
> What about exec-env?

As discussed in #205, this environment will also be used during the build (I haven't added a test for this yet).

**Comment by rgrinberg** [comment]
Well, build commands are also *executed*. There's also `env-vars`, `process-env` perhaps.

**Comment by emillon** [comment]
I just rebased now that env tracking works correctly. I have two questions:
- what does the `context` component of `env_nodes` correspond to? (see initial question)
- how can I add a new field to a record that is only available after a particular syntax version? It seems I'm missing a combinator (I like `env-vars` to be only parsed for syntax >= 1.4 and otherwise evaluate to the empty list).

Thanks!

**Comment by diml** [comment]
> what does the context component of env_nodes correspond to? (see initial question)

It's to represent `dune-workspace` files like this:

```
(context (default (env (flags ...))))
```

i.e. to allow the user to customize individual build contexts.

> how can I add a new field to a record that is only available after a particular syntax version? It seems I'm missing a combinator (I like env-vars to be only parsed for syntax >= 1.4 and otherwise evaluate to the empty list).

Like this:

```ocaml
field "env_vars" ~default:[] (Syntax.since syntax (1, 4) >>> list ...)
```

**Comment by emillon** [comment]
Thanks for the answers @diml. I added some tests for variables set from `(context)` as well. What should be the priority between the two sources?

**Comment by nojb** [comment]
I'm coming late to the discussion, but will this feature allow setting variables such as `PATH` that will themselves be used during the build? (My use-case is adding the path to the C compiler to `PATH` so that it is used by the OCaml compiler).

**Comment by emillon** [comment]
@nojb yes that's the plan. However for PATH there might be some particular handling to do since PATH is also interpreted by dune itself.

**Comment by diml** [comment]
@emillon the context specific variables are more specific, so they should have precedence over the workspace ones

File: src/context.ml [hunk 3]
```diff
@@ -24,6 +24,19 @@ module Env_nodes = struct
...
+      let open Option.O in
+      Option.value
+        ~default:Env.Map.empty
+        (l >>= fun stanza ->
+         Dune_env.Stanza.find stanza ~profile >>| fun {env_vars; _} ->
+         Env.Map.of_list_exn env_vars)
+    in
+    Env.Map.superpose
+      (make_map env_nodes.context)
+      (make_map env_nodes.workspace)
```
Comment (diml) [inline src/context.ml]: So these two lines should be inverted

Comment (emillon) [reply src/context.ml]: Thanks, done

**Comment by rgrinberg** [comment]
Do we need a test case for setting the env-vars in an env stanza of a dune file? Or is that tested elsewhere?

**Comment by emillon** [comment]
Right, that would be nice. Plus, we need some docs and that should be good.

**Comment by rgrinberg** [comment]
Ah yes, the docs are definitely required for this to go in the release. Up to you if you want to do them in a separate PR or not.

**Comment by rgrinberg** [comment]
@emillon could you give this a rebase? I'd like to have a look on top of the up to date code.

**Comment by emillon** [comment]
@rgrinberg I just rebased. I'm changing to use Env.t and add other tests (such as what happens when there are duplicate keys).

**Comment by emillon** [comment]
Setting env vars from dune files is trickier since this relies on the supercontext for the dir -> env_node mapping, so I'll leave that part out of the PR.
